### DISCUSSION TITLE
회원 탈퇴 이후 다시 가입

### DISCUSSION THREAD

**Discussion by soeunnPark** [Kernel360/E2E2-CATCHLINE#50] [General]
회원 탈퇴를 하면 `isMemberDeleted` 값이 `true` 로 변경되어 db에 fetch됩니다.

만약 사용자가 다시 가입하려고 할 때, **중복 이메일 체크**를 하는 과정에서 예외가 발생합니다.

따라서 repository에 다음과 같은 코드를 추가하고, 서비스도 수정했습니다.

```java
@Repository
public interface MemberRepository extends JpaRepository<MemberEntity, Long> {

    // 탈퇴한 회원을 제외하고, 동일한 Email을 가진 회원이 존재하는지 탐색
    Optional<MemberEntity> findByEmailAndIsMemberDeletedFalse(String email);

    // 탈퇴한 회원을 제외하고, 동일한 MemberId를 가진 회원이 존재하는지 탐색
    Optional<MemberEntity> findByMemberIdAndIsMemberDeletedFalse(Long memberId);

}

```

위의 쿼리 메서드는 **중복 이메일 검증**에 사용되며, 아래 쿼리 메서드는 **기존에 존재하는 회원인지 여부를 체크**할 때 사용됩니다.

아래는 수정된 서비스 코드입니다.

```java
// 이미 존재하는 회원인지 체크
    // TODO: 탈퇴한 회원은 제외
    private MemberEntity checkIfMemberPresent(Long memberId) {
        return memberRepository.findByMemberIdAndIsMemberDeletedFalse(memberId)
                .orElseThrow(() -> new IllegalArgumentException("해당하는 사용자가 없습니다."));
    }

    // TODO: 탈퇴한 회원의 이메일이 db에 남아있음. 이들은 제외하고 이메일 중복 체크 해야 함.
    // 이미 존재하는 이메일인지 검증
    private void checkDuplicateEmail(String email) {
        if (memberRepository.findByEmailAndIsMemberDeletedFalse(email).isPresent()) {
            throw new IllegalArgumentException("이미 존재하는 이메일입니다.");
        }
    }

```

기존에 탈퇴한 회원이 로그인, 회원 정보 조회/수정 등이 가능했던 문제는 해결했습니다.

다만, **다시 가입**하려고 할 때 ⚠️**동일한 이메일로는 가입이 여전히 불가능**⚠️합니다.

왜냐하면, 이메일 필드에 `unique` 제약 조건이 걸려 있어서 **이메일 중복 체크 메서드와는 무관**하게, 같은 이메일을 가진 레코드는 db에 추가될 수 없습니다. 

다음과 같은 방법을 생각해봤는데 더 좋은 방법이 있다면 알려주시면 감사하겠습니다!

1. 탈퇴 시 이메일 변경(ex. `carol1234@naver.com` → `deleted_carol1234@naver.com`)
2. `email`, `isMemberDeleted`를 복합 고유 제약 조건 설정
    
    ```java
    @Entity
    @Table(
        name = "member",
        uniqueConstraints = @UniqueConstraint(columnNames = {"email", "isMemberDeleted"})
    )
    public class MemberEntity {
        // 필드 정의
    }
    ```

**Comment by sangwonsheep** [comment]
음... 잠깐 생각해보았을 때 이메일을 null로 변경해서 넣고 싶은데, null을 허용하지 않기 때문에 불가능한 방법이네요.
좋은 방법이 떠오르지 않아서... 이 부분은 멘토님께 질문하면 좋을 것 같습니다!
https://www.notion.so/240802_E2E_2-Catch-Line-2bdd7ae4688844dd97c9f9325f40f7ba
아래 링크에서 질문 추가 부탁드립니다!

**Reply by soeunnPark** [reply]
네 알겠습니다!
